src/core/models.py

```diff
@@ -171,31 +173,46 @@ def password_rotated(self) -> bool:
         return bool(self.relation_data.get("password-rotated", None))
 
     @property
-    def hostname(self) -> str:
+    def hostname(self) -> str | None:
         """The hostname for the unit."""
         return socket.gethostname()
 
     @property
-    def fqdn(self) -> str:
+    def fqdn(self) -> str | None:
         """The Fully Qualified Domain Name for the unit."""
-        return socket.getfqdn()
+        return socket.getfqdn(self.private_ip)
 
     @property
     def private_ip(self) -> str:
         """The IP for the unit."""
-        return socket.gethostbyname(self.hostname)
+        # Either values are okay for us because we are interested on their name
+        # both objects have a .name property.
+        address = None
+        if relation := self.relation:
+            relation_obj = self.model.get_binding(relation.name)
+            address = (
+                str(relation_obj.network.bind_address)
+                if relation_obj and relation_obj.network
+                else None
+            )
+        # Optional fallback, if no binding is found
+        return address or socket.gethostbyname(socket.gethostname())
 
     @property
     def public_ip(self) -> str:
-        result = subprocess.check_output(
-            [
-                "bash",
-                "-c",
-                "ip a | grep global | grep -v 'inet 10.' | cut -d' ' -f6 | cut -d'/' -f1",
```

Comment:
This check will miss other [private IP ranges](https://datatracker.ietf.org/doc/html/rfc1918#section-3).
[Charm-helpers has a lot of facilities to check for IPs straight from interfaces, using netifaces](https://github.com/juju/charm-helpers/blob/1b2d4dc8f8effd79d782241a32a0485af1f01e73/charmhelpers/contrib/network/ip.py#L174-L212). However, that said, I'd still stick with Juju-related information.

Reply:
@skourta, in this case we should consider `netifaces` or any similar lib. However, this is the `public_ip` method. Rarely you will really have a public IP (e.g. 3.4.5.6) assigned to your local interfaces. These will generally be assigned to firewalls, load balancers, etc; any forwarder sitting in a DMZ.